source/portable/mbedtls/core_pkcs11_mbedtls.c

```diff
@@ -4008,6 +4012,176 @@ CK_DECLARE_FUNCTION( CK_RV, C_Sign )( CK_SESSION_HANDLE hSession,
+                                           pxMdInfo,
+                                           PKCS11_USING_HMAC );
+
+        if( lMbedTLSResult != 0 )
+        {
+            LogError( ( "Failed to initialize verify operation. "
+                        "mbedtls_md_setup failed: mbed TLS error = %s : %s.",
+                        mbedtlsHighLevelCodeOrDefault( lMbedTLSResult ),
+                        mbedtlsLowLevelCodeOrDefault( lMbedTLSResult ) ) );
+            prvVerifyInitHMACCleanUp( pxSession );
+            xResult = CKR_KEY_HANDLE_INVALID;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        lMbedTLSResult = mbedtls_md_hmac_starts( &pxSession->xHMACSecretContext,
+                                                 pucKeyData, ulKeyDataLength );
+
+        if( lMbedTLSResult != 0 )
+        {
+            LogError( ( "Failed to initialize verify operation. "
+                        "mbedtls_md_setup failed: mbed TLS error = %s : %s.",
+                        mbedtlsHighLevelCodeOrDefault( lMbedTLSResult ),
+                        mbedtlsLowLevelCodeOrDefault( lMbedTLSResult ) ) );
+            prvVerifyInitHMACCleanUp( pxSession );
+            xResult = CKR_KEY_HANDLE_INVALID;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        pxSession->xHMACKeyHandle = hKey;
+    }
+
+    return xResult;
+}
+
+/**
+ * @brief Helper function for cleaning up a verify operation for an EC or RSA key.
+ * @param[in] pxSession   Pointer to a valid PKCS #11 session.
+ */
+static void prvVerifyInitECRSACleanUp( P11Session_t * pxSession )
+{
+    mbedtls_pk_free( &pxSession->xVerifyKey );
+    pxSession->xVerifyKeyHandle = CK_INVALID_HANDLE;
+}
+
+
+/**
+ * @brief Helper function for initializing a verify operation for an EC or RSA key.
+ * @param[in] pxSession   Pointer to a valid PKCS #11 session.
+ * @param[in] pMechanism  EC/RSA mechanism.
+ * @param[in] hKey        EC/RSA public key handle.
+ * @param[in] pucKeyData        EC/RSA public key data.
+ * @param[in] ulKeyDataLength   EC/RSA public key size.
+ */
+static CK_RV prvVerifyInitECRSAKeys( P11Session_t * pxSession,
+                                     CK_MECHANISM_PTR pMechanism,
+                                     CK_OBJECT_HANDLE hKey,
+                                     CK_BYTE_PTR pucKeyData,
+                                     CK_ULONG ulKeyDataLength )
+{
+    /* See explanation in prvCheckValidSessionAndModule for this exception. */
+    /* coverity[misra_c_2012_rule_10_5_violation] */
+    CK_BBOOL xIsPrivate = ( CK_BBOOL ) CK_TRUE;
+    mbedtls_pk_type_t xKeyType;
+    int32_t lMbedTLSResult = 0;
+    CK_RV xResult = CKR_KEY_HANDLE_INVALID;
+
+    mbedtls_pk_init( &pxSession->xVerifyKey );
+    lMbedTLSResult = mbedtls_pk_parse_public_key( &pxSession->xVerifyKey, pucKeyData, ulKeyDataLength );
+
+    if( 0 == lMbedTLSResult )
+    {
+        pxSession->xVerifyKeyHandle = hKey;
+        xResult = CKR_OK;
+    }
+
+    if( xResult != CKR_OK )
```

Comment:
This is some legacy logic from before I worked on this stack. I agree with you that we should be more explicit, and not try to parse a private key. 

Unfortunately I don't have the context for the _why_ when it comes to this. Partially I recall some weirdness with how the keys were being written to the flash, of the private and public key sharing a location but the offset being parsed. 

Since then though I have changed the logic quite a bit, and separated the public key and the private key in the flash.

What I propose here is we leave this as is, and put it on the backlog. I think this change should be straightforward, but I would like to isolate it so we can properly test it (Might cause failures in AFR, I think the windows and POSIX port will be fine).

Reply:
Backlog sounds fine.

Reply:
Added a comment